crates/ruff_linter/src/rules/pylint/rules/self_cls_assignment.rs

```diff
@@ -0,0 +1,104 @@
+use ruff_python_ast::{self as ast, Expr};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_semantic::{analyze::function_type, ScopeKind};
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for assignment of `self` and `cls` in methods.
+///
+/// ## Why is this bad?
+/// The identifiers `self` and `cls` are conventional in Python for
+/// the first argument of instance methods and class methods, respectively.
+///
+#[violation]
+pub struct SelfClsAssignment {
+    keyword: String,
+}
+
+impl Violation for SelfClsAssignment {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let SelfClsAssignment { keyword } = self;
+        format!("Assignment of variable `{keyword}`")
+    }
+}
+
+/// PLW0127
+pub(crate) fn self_cls_assignment(checker: &mut Checker, target: &Expr) {
```

Comment:
As suggested by Micha, can you rename this to `self_or_cls_assignment`?
```suggestion
pub(crate) fn self_or_cls_assignment(checker: &mut Checker, target: &Expr) {
```